### DISCUSSION TITLE
AttributeError: partially initialized module 'jax' has no attribute 'version' (most likely due to a circular import)

### DISCUSSION THREAD

**Discussion by lexie-demarco** [jax-ml/jax#14036] [Q&A]
I am trying to use jax.numpy and am getting the error "AttributeError: partially initialized module 'jax' has no attribute 'version' (most likely due to a circular import)" Nothing I have tried worked and would appreciate any help.
<img width="882" alt="Screen Shot 2023-01-17 at 1 49 40 PM" src="https://user-images.githubusercontent.com/122923738/212985706-510d4b00-2602-4405-928d-b62922b118ee.png">

**Comment by jakevdp** [comment]
This typically means you have some kind of mismatch between the version of `jax` and the version of `jaxlib` in your environment. You can check those by running:
```
pip list | grep jax
```
To fix it, try updating to the most recent jax/jaxlib version:
```
pip install -U jax jaxlib
```
(If you need GPU support, see the [install guide](http://go/jax-github#installation))

Also be certain to restart your Colab runtime any time you update packges, otherwise cached package versions may be used and could cause the error you're seeing.

**Reply by lexie-demarco** [reply]
I tried both of these and confirmed that they are the same version and also updated to the most recent jax/jaxlib version but I am still getting the same error. Do you know of anything else that could cause this problem? Thanks!

**Reply by jakevdp** [reply]
Did you restart your colab session after installing new versions in the runtime?

**Reply by coolalu** [reply]
Hi there,

I tried to resolve the issue by updating jax/jaxlib but it's not working - 
![Screenshot 2023-04-12 at 6 38 51 PM](https://user-images.githubusercontent.com/129926237/231601557-0815111d-7496-41fa-aeb9-96c35015e642.png)

Both are updated as above but I am still getting the error. Do you know other ways that could help? Thank you!

**Reply by jakevdp** [reply]
Did you restart your notebook session after installing new versions in the runtime?

**Reply by coolalu** [reply]
Hi yes I restarted my notebook but it returned another error - https://github.com/google/lightweight_mmm/issues/197
![image](https://user-images.githubusercontent.com/129926237/232127268-85c29c46-34d1-4257-b498-170906445d0c.png)
After I tried update to the most recent jax/jaxlib version, the first error about incompatible system returned.
Do you happen to know how to get rid of both errors? Thanks!

**Reply by jakevdp** [reply]
Regarding AVX instructions, see this thread: https://github.com/google/jax/issues/9580

If you're on a non-AVX system I believe your only recourse is to build jaxlib yourself; see https://jax.readthedocs.io/en/latest/developer.html#building-from-source for information on this.

**Reply by GeorgiadeG** [reply]
> Did you restart your notebook session after installing new versions in the runtime?

I had the same error, update both jax and jaxlib exactly the same as shown by OP, restart my notebook and the error doesn't change. Any ideas of what else might be wrong?

**Reply by jakevdp** [reply]
The error here indicates that the `jax` and `jaxlib` versions are out-of-sync. This can be caused by having mismatched versions installed, or by updating versions without restarting your jupyter or colab runtime.

Are you certain you're importing the jax/jaxlib versions you think you are? Especially when working in Jupyter notebooks, it's possible that you're installing the packages in a different environment than your notebook runtime. There's a decent writeup of this issue here: https://stackoverflow.com/questions/39007571/running-jupyter-with-multiple-python-and-ipython-paths/39022003#39022003

**Reply by dmilakovic** [reply]
Hi. 
I have the same error. 

It seems to me that the issue stems from the fact that `jax.version` is not defined before provided to function `check_jaxlib_version` in line 74 of the file `jax/_src/lib/__init__.py`, As a result, the function `check_jaxlib_version` fails to run at all.

Reading previous responses, I confirmed that the `jax` and `jaxlib` versions match so that is unlikely to be the culprit. Yet, the problem persists. Other information that may be useful:
jax                           0.4.10
jaxlib                        0.4.10
jaxopt                        0.6

Python 3.9.12
MacOS 12.2.1

I am not using jupyter or other collaborative tools.

Many thanks!

**Reply by jakevdp** [reply]
The only time I've seen this error is when JAX and jaxlib are out of sync. If you believe you have a different issue, I'd suggest opening a new issue with more details. Long comment threads on old discussions tend to fall under the radar.

**Reply by LiamLau23** [reply]
but why i run the demand "pip list | grep jax" is wrong?

**Comment by Gaurav17Joshi** [comment]
I was getting the same issue, and I uninstalled and used conda forge to reinstall jax and jaxlib `conda install -c conda-forge jax`, `conda install -c conda-forge jaxlib`. My versions for both were 0.4.12.

**Reply by ingomueller-net** [reply]
I also ran into this issue after installing the wheel I built for `jaxlib` and then installing `jax` with `pip install -e .`. When I installed `jax` with `pip install .` (i.e., not as editable), then it the error went away.

**Comment by hepnerb** [comment]
Switching from Anaconda on Windows to WSL2 solved the issues for me. Not a full fix but a quick one.

**Comment by rajkk1** [comment]
I had the same problem, but then I uninstalled jax and jaxlib, and re-installed GPU version of jax with `pip install -U "jax[cuda12_pip]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html` (and didn't reinstall jaxlib) and that fixed the problem somehow 🤷🏾

**Reply by sinodine** [reply]
I tried that, but it didn't work for me.
What worked was:
- unistall jax and jaxlib
- `pip install -U "jax[cuda12_pip]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html`
- `pip install -U "jax[cpu]"`    
The last command installed the jax 0.4.6 installed with the second line and reinstalled on it's own jax-0.4.28 and jaxlib-0.4.28.
